File: 496. Next Greater Element I/496. Next Greater Element I.java
public class Solution {
    public int[] nextGreaterElement(int[] findNums, int[] nums) {
        if(findNums == null ||  nums == null || 
           findNums.length == 0 || nums.length == 0 || 
           findNums.length > nums.length) return new int[0];

        int m = findNums.length;
        int n = nums.length;
        int[] result = new int[m];
        Map<Integer, Integer> map = new HashMap<Integer, Integer>();

        for(int j = 0; j < n; ++j){
            map.put(nums[j], j);
        }
        for(int i = 0; i < m; ++i){
            int j = map.get(findNums[i]);
            for(; j < n; ++j){
                if(nums[j] > findNums[i]) break;
            }
            result[i] = j < n ? nums[j] : -1;
        }
        return result;
    }
}


//解法二 ，leetcode牛方法

'''
重点观察： 假设我们有一个递减的序列，后跟一个更大的数字 例如[5,4,3,2,1,6]，则数字6越大，序列中所有先前数字的下一个更大元素 
我们使用一个堆栈来保持一个递减的子序列，每当我们看到一个数字x大于stack.peek（）时我们弹出所有小于x的元素，对于所有弹出的元素，
它们的下一个更大的元素是x 例如[9,8,7,3,2,1,6] 堆栈首先包含[9,8,7,3,2,1]，然后我们看到6大于1所以我们弹出1 2 3，
其下一个更大的元素应该是6
'''
public int[] nextGreaterElement(int[] findNums, int[] nums) {
    Map<Integer, Integer> map = new HashMap<>(); // map from x to next greater element of x
    Stack<Integer> stack = new Stack<>();
    for (int num : nums) {
        while (!stack.isEmpty() && stack.peek() < num)
            map.put(stack.pop(), num);
        stack.push(num);
    }   
    for (int i = 0; i < findNums.length; i++)
        findNums[i] = map.getOrDefault(findNums[i], -1);
    return findNums;
}

File: 496. Next Greater Element I/496. Next Greater Element I.cpp
/*
解题思路：

解法一：暴力解法，遍历nums1，对其中的每个数，在nums2中找到该数，然后在nums2中该数的右边找第一个较大的数。
*/


public int[] nextGreaterElement(int[] findNums, int[] nums) {
		int findLen = findNums.length;
		int numsLen = nums.length;
		int i, j, k;
		int[] res = new int[findLen];
		for (i = 0; i < findLen; i++) {
			for (j = 0; j < numsLen; j++) {
				if (findNums[i] == nums[j]) {
					for (k = j + 1; k < numsLen; k++) {
						if (nums[k] > findNums[i]) {
							res[i] = nums[k];
							break;
						}
					}
					if (k == numsLen)
						res[i] = -1;
					break;
				}
			}
		}
		return res;
	}


/* 
解法二：首先确定nums2中每一个数和右边第一个较大数的关系。方法为维护一个栈和一个map，从尾到头遍历nums2，对于每一个数，
如果栈顶元素小于该数，则栈顶元素出栈，直到找到大于该数的栈顶元素，此时将<该数，栈顶元素>放入map中，然后将该数也入栈；如果栈空了，
则将<该数，-1>放入map中，同样将该数入栈。确定好这样一个关系后，就可以直接遍历nums1，对每一个元素直接在map中执行get，
就可以确定nums2中下一个较大元素的值了。

*/

public int[] nextGreaterElement(int[] findNums, int[] nums) {
        int findLen = findNums.length;
		int numsLen = nums.length;
		int[] res = new int[findLen];
		if (findLen == 0 || numsLen == 0 || findLen > numsLen)
			return res;
		Map<Integer, Integer> map = new HashMap<>();
		int[] stack = new int[numsLen];
		int index = -1;
		int i, j;
		int curMax = nums[numsLen - 1];
		map.put(curMax, -1);
		for (i = numsLen - 1; i >= 0; i--) {
			while (index >= 0) {
				if (nums[i] > stack[index])
					--index;
				else {
					map.put(nums[i], stack[index]);
					stack[++index] = nums[i];
					break;
				}
			}
			if (index < 0) {
				map.put(nums[i], -1);
				stack[++index] = nums[i];
			}
		}
		for (j = 0; j < findLen; j++) {
			res[j] = map.get(findNums[j]);
		}
		return res;
    }


//方法三  多年未见大佬
class Solution {
public:
    vector<int> nextGreaterElement(vector<int>& findNums, vector<int>& nums) {
        map<int, int> mp;
        int cc = 1;
        int n = findNums.size(), m = nums.size();
        for (int i = 0; i < n; i++) {
            mp[findNums[i]] = cc++;
        }
        vector<int> ans(n, -1);
        stack<int> s;
        for (int i = 0; i < m; i++) {
            while (!s.empty() && nums[s.top()] < nums[i]) {
                int tmp = mp[nums[s.top()]];
                if (tmp) ans[tmp - 1] = nums[i];
                s.pop();
            }
            s.push(i);
        }
        while (!s.empty()) s.pop();
        return ans;
    }
};
